### PR TITLE
Added HoloInvertModel()

### PR DESCRIPTION
**Pull request by unknao**
Added an e2 function for inverting models.

### CODE REVIEW

**Comment by Fasteroid** [comment]
yes please

File: lua/entities/gmod_wire_hologram.lua [hunk 1]
```diff
@@ -120,14 +120,20 @@ if CLIENT then
...
 			render.SuppressEngineLighting(true)
 			self:DrawModel()
 			render.SuppressEngineLighting(false)
 		else
 			self:DrawModel()
 		end
+		
+		if self:GetNWBool("invert_model") then
+			render.CullMode(0)
+		end
```
Comment (thegrb93) [inline lua/entities/gmod_wire_hologram.lua:136]: does this work?
```lua
local oldcull = render.CullMode(self:GetNWBool("invert_model", 0))
...
render.CullMode(oldcull)
```

File: lua/entities/gmod_wire_expression2/core/hologram.lua [hunk 2]
```diff
@@ -1057,6 +1057,14 @@ e2function void holoDisableShading( index, disable )
 	Holo.ent:SetNWBool( "disable_shading", disable == 1 )
 end
 
+e2function void holoInvertModel( index, enable )
+	local Holo = CheckIndex(self, index)
+	if not Holo then return end
+	local enable = math.clamp(math.ceil(enable), 0, 1)
```
Comment (Vurv78) [inline lua/entities/gmod_wire_expression2/core/hologram.lua:1063]: This should be inlined (and use `~= 0`, not this logic)

File: lua/entities/gmod_wire_expression2/core/hologram.lua [hunk 3]
```diff
@@ -1057,6 +1057,14 @@ e2function void holoDisableShading( index, disable )
...
+	
+	Holo.ent:SetNWBool( "invert_model", enable)
```
Comment (Vurv78) [inline lua/entities/gmod_wire_expression2/core/hologram.lua:1065]: Random space before arguments

File: lua/wire/client/e2descriptions.lua [hunk 4]
```diff
@@ -1205,6 +1205,7 @@ E2Helper.Descriptions["holoClip(nvve)"] = "Defines a plane used to clip a hologr
 E2Helper.Descriptions["holoClipEnabled(nn)"] = "Enables / disables clipping for a hologram with specified index"
 E2Helper.Descriptions["holoClipEnabled(nnn)"] = "Enables / disables clipping for a hologram with specified index. Clip index is for use with multiple clipping planes"
 E2Helper.Descriptions["holoClipsAvailable()"] = "Returns the maximum number of clipping planes allowed per hologram"
+E2Helper.Descriptions["holoInvertModel(nn)"] = "If higher than 0, inverts the model of the hologram"
```
Comment (Vurv78) [inline lua/wire/client/e2descriptions.lua:1208]: If not 0, with above review

**Review by Vurv78 — CHANGES_REQUESTED** [review]

File: lua/entities/gmod_wire_hologram.lua (hunk 1 above)
Comment (unknao) [reply lua/entities/gmod_wire_hologram.lua:136]: render.CullMode doesn't return anything.

**Review by unknao — COMMENTED** [review]
Made changes according to feedback.

**Review by thegrb93 — APPROVED** [review]

File: lua/entities/gmod_wire_expression2/core/hologram.lua [hunk 5]
```diff
@@ -1057,6 +1057,13 @@ e2function void holoDisableShading( index, disable )
 	Holo.ent:SetNWBool( "disable_shading", disable == 1 )
 end
 
+e2function void holoInvertModel( index, enable )
+	local Holo = CheckIndex(self, index)
+	if not Holo then return end
+	
+	Holo.ent:SetNWBool( "invert_model", enable ~= 0 and 1 or 0 )
```
Comment (Vurv78) [inline lua/entities/gmod_wire_expression2/core/hologram.lua:1064]: Shouldn't be doing `and 1 or 0`, I don't think SetNWBool accepts numbers.

File: lua/entities/gmod_wire_hologram.lua [hunk 6]
```diff
@@ -120,14 +120,20 @@ if CLIENT then
 		end
 
 		self:SetupClipping()
-
+		
+		render.CullMode(self:GetNWBool("invert_model", 0))
```
Comment (Vurv78) [inline lua/entities/gmod_wire_hologram.lua:124]: Can you store the result of `self:GetNWBool` as a variable so it isn't called twice?

File: lua/entities/gmod_wire_hologram.lua [hunk 7]
```diff
@@ -120,14 +120,20 @@ if CLIENT then
...
+		
 		if self:GetNWBool("disable_shading") then
 			render.SuppressEngineLighting(true)
 			self:DrawModel()
 			render.SuppressEngineLighting(false)
 		else
 			self:DrawModel()
 		end
+		
+		if self:GetNWBool("invert_model") then
```
Comment (Vurv78) [inline lua/entities/gmod_wire_hologram.lua:134]: This `if` is pointless right now assuming `GetNWBool` would return a number (which is always truthy) from when you `SetNWBool`.

File: lua/entities/gmod_wire_hologram.lua (hunk 6 above)
Comment (Vurv78) [reply lua/entities/gmod_wire_hologram.lua:124]: And if you make the change above would need to do `inverted and 1 or 0`, (or maybe not, if lua can coerce a boolean into a number in the `render.CullMode` C function, if you want to try that out)

**Review by Vurv78 — CHANGES_REQUESTED** [review]

**Comment by unknao** [comment]
Actually the SetNWBool approach works as is, but i forgot about storing the variable and the fact that SetNWInt exists so ill change that.

**Comment by thegrb93** [comment]
That's true, forgot you could do that. Any reason to not just do that?

**Comment by unknao** [comment]
Inverting the model allows you to mirror props that only have no mirrored counterparts (like gibs).

image
Something like this cannot be made without the means to invert models.

**Comment by Fasteroid** [comment]
> Inverting the model allows you to mirror props that only have no mirrored counterparts (like gibs).
Implementing cel-shading with inverted holograms would also be more flexible since there'd no longer be a requirement for model symmetry about one axis.

**Comment by thegrb93** [comment]
Had a feeling it had something to do with symmetry, so yeah I guess it can be added.

**Comment by Fasteroid** [comment]
hot

**Pull request merged**